keras/utils/data_utils.py

```diff
@@ -186,19 +186,21 @@ def get_file(fname,
 
     if download:
         print('Downloading data from', origin)
-        progbar = None
 
-        def dl_progress(count, block_size, total_size, progbar=None):
-            if progbar is None:
-                progbar = Progbar(total_size)
+        # Closures: Use a dictionary workaround To support python2,
+        # since `nonlocal` is only support in python3.
```

Comment:
I could use some help on clarifying this comment. Would adding a link to an [article ](https://technotroph.wordpress.com/2012/10/01/python-closures-and-the-python-2-7-nonlocal-solution/)which explains this technique to the job?
```
# Closures: Use a dictionary workaround To support python2,
# since `nonlocal` is only support in python3.
# For more details: https://goo.gl/JOnlnH
```
I was trying to avoid putting an essay in the comment, but here's the long version.

When you reference a variable in an expression, the Python interpreter will traverse the scope to resolve the reference in this order:

1. The current function’s scope (ex: `dl_progress`)
1. Any enclosing scopes (ex: `get_file`)
1. The scope of the module that contains the code (also called the `global` scope)
1. The built-in scope (that contains functions like `len` and `str`)

The original implementation put the variable `progbar` inside the `global` scope (from the point of view of the `dl_progress` function): #https://github.com/fchollet/keras/commit/1ebeff8ee304833a3df421d0de3b9f9480570eb9#diff-ba9d38600a2df565e5ae8757eb2b1b35R55
```
def get_file(fname, origin):
    global progbar
    progbar = None

    def dl_progress(count, block_size, total_size):
        global progbar
        if progbar is None:
            progbar = Progbar(total_size)
        else:
            progbar.update(count*block_size)
            
    urlretrieve(origin, fpath, dl_progress)
```
This code was perfectly functional, but some view the use of globals as a sign of bad form, which I assume is the reason the code was changed in the commit titled "Further style fixes.": #https://github.com/fchollet/keras/commit/7aa3114d9ff1cfbeb4128934fec05878fb7cf633#diff-ba9d38600a2df565e5ae8757eb2b1b35L87
This commit put `progbar` inside the function's scope (again, from the point of view of the `dl_progress` function), and is the current code we are using.
```
def get_file(fname, origin):
    progbar = None

    def dl_progress(count, block_size, total_size, progbar=None):
        if progbar is None:
            progbar = Progbar(total_size)
        else:
            progbar.update(count*block_size)
            
    urlretrieve(origin, fpath, functools.partial(dl_progress, progbar=progbar))
```
The only problem with this code is that it does not work. It would work if progbar is passed by reference, but Python is neither pass by value nor pass by reference, instead it is a PASS-BY-OBJECT-REFERENCE programming language (Object references are passed by value). THE VARIABLE IS NOT THE OBJECT. The following two examples should help with understanding
```
def append_one(li):
    li.append(1)
l = [0]
append_one(l)
print l
```
As expected, the output of the above program will be:
```[0, 1]```
```
def append_one(li):
    li = [0, 1]
l = [0]
append_one(l)
print l
```
If the variable li is reassigned in the function, then li will point to a seperate object in memory, which is exactly what is happening in the current `dl_progress`. 
Thus, the output of the above program will be:
`[0]`

No worries, there is a [solution](https://en.wikipedia.org/wiki/Closure_(computer_programming)) to this problem called closures, by putting `progbar` inside the enclosing function's scope (again, from the point of view of the dl_progress function), which is `get_file`'s scope:
```
def get_file(fname, origin):
    progbar = None

    def dl_progress(count, block_size, total_size):
        if progbar is None:
            progbar = Progbar(total_size)
        else:
            progbar.update(count*block_size)
            
    urlretrieve(origin, fpath, dl_progress)
```
Except that the above code will spit out Errors and not run at all, because `progbar` is reasigned in `dl_progress` at the line `progbar = Progbar(total_size)`, this line makes `progbar` a local variable inside `dl_progress`, which has a higher precedence then the enclosing `progbar` defined in `get_file`. The ideal fix for this problem is to use the `nonlocal` keyword:
```
def get_file(fname, origin):
    progbar = None

    def dl_progress(count, block_size, total_size):
        nonlocal progbar
        if progbar is None:
            progbar = Progbar(total_size)
        else:
            progbar.update(count*block_size)
            
    urlretrieve(origin, fpath, dl_progress)
```
This instructs `dl_progress` to use the `get_file` version or `progbar`, and the code is once again perfectly functional, the only problem is, the `nonlocal` keyword does not exist in Python2, not in __future__, and not in six.

Finally, we arrive at version I used in this PR
```
def get_file(fname, origin):
    enclosed = {'progbar': None}

    def dl_progress(count, block_size, total_size):
        if enclosed['progbar'] is None:
            enclosed['progbar'] = Progbar(total_size)
        else:
            enclosed['progbar'].update(count*block_size)
            
    urlretrieve(origin, fpath, dl_progress)
```
The `enclosed` variable is a dictionary, since the `enclosed` variable is not redefined in `dl_progress`, `enclosed` does not exist in `dl_progress`'s local scope, hence it will find it in the enclosed scope (`get_file`'s scope)

So again, I could use some help on clarifying this comment (English is my fourth language).